File: python/circularLinkedList.py
class node:
    def __init__(self,data,next=None):
        self.data=data
        self.next=next
    
class circular:
    #constructor
    def __init__(self,data=None):
        self.head=node(data)
        self.head.next=self.head      
    
    #modifiers
    def insertAtTail(self,data):
        it=self.head
        newNode=node(data,self.head)

        while True: 
            if it.next==self.head:
                break
            it=it.next
        it.next=newNode

    def insertAtHead(self,data):
        newNode=node(data,self.head)
        it=self.head
        while True:
            if it.next==self.head:
                break
            it=it.next
        it.next=newNode
        self.head=newNode
        return

    def insertAtEle(self,key,data):
        newNode=node(data)
        it=self.head
        while True:
            if it.next==self.head or it.data==key:
                break
            it=it.next
        newNode.next=it.next
        it.next=newNode
        return
    
    def deleteAtTail(self):
        it=self.head
        if self.head.next==self.head:
            self.deleteAtHead()

        while True:
            if it.next.next==self.head:
                toDel=it.next
                it.next=self.head
                del toDel
                return
            it=it.next

    def deleteAtHead(self):
        it=self.head
        toDel=self.head
        if it is None:
            print("Empty list")
            return
        
        if it.next==self.head:
            self.head=None
            return
        
        while True:
            if it.next==self.head:
                break
            it=it.next
        it.next=self.head.next
        self.head=self.head.next
        del toDel
        return
    
    #accessing/traversal
    def printList(self):
        it=self.head
        if self.head is None:
            print(self.head)
            return
        while True:
            print(it.data,end='->')
            it=it.next
            if it==self.head:
                break
        print(self.head.data)

if __name__=='__main__':
    root=circular(int(input("Enter the starting value -> ")))
    root.printList()
    print("After inserting 4 at tail : ",end="")
    root.insertAtTail(4)
    root.printList()
    print("After inserting 2 at head :",end=" ")
    root.insertAtHead(2)
    root.printList()
    print("After inserting 6 after 2 :",end=" ")
    root.insertAtEle(2,6)
    root.printList()
    print("After deleting at tail :",end=" ")
    root.deleteAtTail()
    root.printList()
    print("After deleting element at head :",end=" ")
    root.deleteAtHead()
    root.printList()

File: cpp/circularLinkedList.cpp
#include <bits/stdc++.h>
#define ll long long
#define ld long double
#define forPrintLoop(arr,n) for(int i=0;i<n;i++) std::cout << arr[i] << ' ';
#define forInputLoop(arr,n) for(int i=0;i<n;i++) std::cin >> arr[i]
#define space ' '
#define newline '\n'
using namespace std;

class node{
    public:
        node * next;
        int data;
        node(int data,node *next=NULL):data{data},next{next}{}
};

void insertAtHead(node *&head,int data){
    node *newHead=new node(data,head);
    if(head==NULL){
        head=newHead;
        newHead->next=head;
        return;
    }   

    auto it=head;
    while(true){
        it=it->next;
        if(it->next==head){
            break;
        }
    }  

    it->next=newHead;
    head=newHead;
    return;  
}

void insertAtTail(node *&head,int data){
    auto insert=new node(data,head);
    if(head==NULL){
        head=insert;
        insert->next=head;
        return;
    }
    auto it=head;
    while(true){
        if(it->next==head){
            break;
        }
        it=it->next;
    }
    it->next=insert;
    return;
}
ostream& operator <<(ostream &os,node *head){
    auto it=head;
    cout << "(start)";
    while(true){
        os<<it->data << "->";
        it=it->next;
        if(it==head){
            break;
        }
    }
    os << it->data << "(end)" << newline;
    return os;
}

int main()
{   
    node *head=NULL;
    insertAtHead(head,3);
    insertAtHead(head,2);
    cout << "After inserting 3 and 2 at head -> " << newline;
    cout << head;
    insertAtTail(head,4);
    insertAtTail(head,5);
    cout << "After inserting 4 and 5 at tail ->" << newline;
    cout << head;
    return 0;
}